File: frontend/src/services/notifications.ts
// src/services/notifications.ts
import axios from "axios";

const API_URL = import.meta.env.VITE_API_URL || "http://localhost:5000";
const authHeader = () => {
  const token = localStorage.getItem("access_token");
  return token ? { Authorization: `Bearer ${token}` } : {};
};

export type Notification = {
  id: number;
  user_id: number;
  kind: string;
  title: string;
  body: string;
  ref_type?: string | null;
  ref_id?: number | null;
  scheduled_for?: string | null;
  delivered_at?: string | null;
  read_at?: string | null;
  created_at: string;
  updated_at: string;
};

export async function scanDue(opts?: { soon_hours?: number; overdue_days?: number }) {
  const params: any = {};
  if (opts?.soon_hours) params.soon_hours = opts.soon_hours;
  if (opts?.overdue_days) params.overdue_days = opts.overdue_days;
  const res = await axios.get(`${API_URL}/api/notifications/scan-due`, { headers: authHeader(), params });
  return res.data as { created: number };
}

export async function fetchUnread() {
  const res = await axios.get(`${API_URL}/api/notifications/unread`, { headers: authHeader() });
  return res.data as { items: Notification[]; count: number };
}

export async function markRead(ids: number[]) {
  const res = await axios.post(
    `${API_URL}/api/notifications/mark-read`,
    { ids },
    { headers: { ...authHeader(), "Content-Type": "application/json" } }
  );
  return res.data as { updated: number };
}

File: backend/app/routes/notifications.py
from datetime import datetime
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity

from app.extensions import db
from app.models.notification import Notification
from app.services.notify import scan_user_tasks_for_reminders

notifications_bp = Blueprint("notifications", __name__, url_prefix="/api/notifications")


@notifications_bp.route("/scan-due", methods=["POST","GET"])  # allow GET for easy Postman
@jwt_required()
def scan_due():
    uid = get_jwt_identity()
    soon_h = int(request.args.get("soon_hours", 24))
    over_d = int(request.args.get("overdue_days", 7))
    out = scan_user_tasks_for_reminders(uid, window_soon_hours=soon_h, overdue_window_days=over_d)
    return out, 200


@notifications_bp.route("/unread", methods=["GET"])  # unread for user
@jwt_required()
def unread():
    uid = get_jwt_identity()
    items = Notification.query.filter_by(user_id=uid).filter(Notification.read_at == None).order_by(Notification.created_at.desc()).limit(50).all()  # noqa: E711
    return {"items": [n.to_dict() for n in items], "count": len(items)}, 200


@notifications_bp.route("/mark-read", methods=["POST"])  # mark one or many
@jwt_required()
def mark_read():
    uid = get_jwt_identity()
    data = request.get_json(silent=True) or {}
    ids = data.get("ids") or []
    now = datetime.utcnow()
    q = Notification.query.filter(Notification.user_id == uid, Notification.id.in_(ids))
    updated = 0
    for n in q.all():
        if n.read_at is None:
            n.read_at = now
            updated += 1
    db.session.commit()
    return {"updated": updated}, 200


@notifications_bp.route("/all", methods=["GET"])  # paginated history
@jwt_required()
def list_all():
    uid = get_jwt_identity()
    page = max(1, int(request.args.get("page", 1)))
    page_size = min(100, max(1, int(request.args.get("page_size", 20))))
    q = Notification.query.filter_by(user_id=uid).order_by(Notification.created_at.desc())
    items = q.paginate(page=page, per_page=page_size, error_out=False)
    return {
        "items": [n.to_dict() for n in items.items],
        "page": page,
        "page_size": page_size,
        "total": items.total,
        "pages": items.pages,
    }, 200
